Libraries/DirectMLX.h

```diff
@@ -2531,9 +2531,47 @@ namespace dml
         return output;
     }
 
-    // 
-    // TODO: GatherND (DML_OPERATOR_GATHER_ND1)
-    // 
+    inline Expression GatherND(
+        Expression input,
+        Expression indices,
+        uint32_t inputDimensionCount,
+        uint32_t indicesDimensionCount,
+        uint32_t batchDimensionCount)
+    {
+        detail::GraphBuilder* builder = input.Impl()->GetGraphBuilder();
+        TensorDesc inputTensor = input.Impl()->GetOutputDesc();
+        TensorDesc indicesTensor = indices.Impl()->GetOutputDesc();
+
+        uint32_t dimensionCount = static_cast<uint32_t>(inputTensor.sizes.size());
+        assert(indicesTensor.sizes.size() == dimensionCount);
+        assert(inputDimensionCount <= dimensionCount);
+        assert(indicesDimensionCount <= dimensionCount);
+        assert(batchDimensionCount <= dimensionCount);
+
+        uint32_t numberOfCoordinatesPerIndex = indicesTensor.sizes.back();
+        uint32_t numberOfOutputDimensionsFromInput = inputDimensionCount - batchDimensionCount - numberOfCoordinatesPerIndex;
```

Comment:
Should we also `assert` that `numberOfOutputDimensionsFromInput` is <= `inputDimensionCount `? Because if it wraps around due to underflow, then that's problematic later on line 2555.